In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


** THIS PHASE BEGINS WITH THE INVESTIGATION OF THE PROCESSED DATA IN PHASE 5 ABOVE AND A BRIEF CONCLUSION OF MY FINDINGS **

In [ ]:
# Definnig the phase 5 folder
import pandas as pd
import os

processed_path = "/content/drive/MyDrive/Meridian Retail Group - Customer Churn Capstone /Data/Processed Data"

print("Files in Processed Data folder:")
print(os.listdir(processed_path))

Files in Processed Data folder:
['products_processed.csv', 'support_tickets_processed.csv', 'customers_processed.csv', 'order_items_processed.csv', 'orders_processed.csv']


In [ ]:
# Loading all five Dataset
customers = pd.read_csv(
    f"{processed_path}/customers_processed.csv"
)

products = pd.read_csv(
    f"{processed_path}/products_processed.csv"
)

orders = pd.read_csv(
    f"{processed_path}/orders_processed.csv"
)

order_items = pd.read_csv(
    f"{processed_path}/order_items_processed.csv"
)
support_tickets = pd.read_csv(
    f"{processed_path}/support_tickets_processed.csv"
)

In [ ]:
# Confirming that it loaded
print("Customers:", customers.shape)
print("Products:", products.shape)
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Support Tickets:", support_tickets.shape)

Customers: (6090, 23)
Products: (220, 10)
Orders: (28297, 9)
Order Items: (73505, 5)
Support Tickets: (5200, 8)


In [ ]:
# Initial Data Inspection
print("========== CUSTOMERS ==========")
display(customers.head())

print("\n========== PRODUCTS ==========")
display(products.head())

print("\n========== ORDERS ==========")
display(orders.head())

print("\n========== ORDER ITEMS ==========")
display(order_items.head())
print("\n========== SUPPORT TICKETS ==========")
display(support_tickets.head())

========== CUSTOMERS ==========


,customer_id,first_name,last_name,email,phone,gender,date_of_birth,signup_date,state,city,...,referral_source,Age,Age_Group,Order_Count,Total_Revenue,Customer_Segment,Revenue_ZScore,Revenue_Outlier,Ticket_Count,Average_Satisfaction
0,CUST002565,Karen,Castro,karen.castro168@gmail.com,784-499-9916x95359,F,1957-07-03,2026-05-28,FL,Jeffreybury,...,Email Campaign,69.0,55+,8.0,13346.48,High-Value Customer,1.415829,No,2.0,4.0
1,CUST004275,Tonya,Webster,tonya.webster780@gmail.com,3985078721,M,1966-12-07,2022-12-23,Massachusetts,Michaelfurt,...,Search Engine,59.0,55+,0.0,0.00,No Orders,-1.638622,No,1.0,3.0
2,CUST003045,Alexis,Thomas,alexis.thomas522@gmail.com,5624741570,Other,1998-06-14,2022-07-10,Wisconsin,East Mark,...,TV Ad,28.0,25-34,3.0,8742.62,Repeat Customer,0.362198,No,1.0,5.0
3,CUST005168,Danielle,Goodman,danielle.goodman202@gmail.com,505.988.6915x03091,M,1975-08-30,NaN,in,Ramosberg,...,Friend Referral,51.0,45-54,4.0,9395.18,Repeat Customer,0.511542,No,1.0,3.0
4,CUST003986,Daniel,Bowman,daniel.bowman295@yahoo.com,(988)305-0745x11524,F,1999-09-15,2023-07-05,TX,Walkerside,...,NaN,26.0,25-34,5.0,6696.85,Repeat Customer,-0.105993,No,1.0,4.0



========== PRODUCTS ==========


,product_id,product_name,category,sub_category,brand,unit_cost,unit_price,launch_date,is_active,Price_Level
0,PROD1000,Cadence Wellness Classic,Beauty & Personal Care,Wellness,Cadence,152.32,415.24,2023-09-08,Yes,High
1,PROD1001,Northfield Cookware Pro,Home & Kitchen,Cookware,Northfield,293.87,645.92,2025-11-10,Yes,High
2,PROD1002,Pinecrest Accessories Essential,Apparel,Accessories,Pinecrest,65.78,NaN,2026-03-13,Yes,High
3,PROD1003,Vertex Camera Essential,Electronics,Cameras,Vertex,27.00,70.18,2023-02-06,Yes,Low
4,PROD1004,Lumen Wellness Plus,Beauty & Personal Care,Wellness,Lumen,242.04,571.72,2025-04-04,Yes,High



========== ORDERS ==========


,order_id,customer_id,order_date,channel,order_status,payment_method,shipping_cost,discount_pct,total_amount
0,ORD118371,CUST005211,2026-02-14,Online,Completed,Gift Card,0.00,5,1098.00
1,ORD103955,CUST004689,2021-06-07,Mobile App,Completed,PayPal,0.00,0,349.47
2,ORD102242,CUST006035,2023-06-01,Online,Completed,Debit Card,12.99,15,981.00
3,ORD125665,CUST002265,2026-04-05,Mobile App,Completed,Debit Card,0.00,0,120.61
4,ORD108154,CUST000471,2022-06-28,Mobile App,Completed,Debit Card,5.99,0,225.89



========== ORDER ITEMS ==========


,order_item_id,order_id,product_id,quantity,unit_price_at_purchase
0,OI562911,ORD116225,PROD1021,3,45.21
1,OI506869,ORD102266,PROD1124,2,982.99
2,OI529520,ORD100873,PROD1129,1,326.95
3,OI545253,ORD126873,PROD1097,2,247.92
4,OI526160,ORD105507,PROD1025,1,940.93



========== SUPPORT TICKETS ==========


,ticket_id,customer_id,ticket_date,issue_category,channel,resolution_time_hours,satisfaction_score,ticket_text
0,TIX100000,CUST004178,2026-04-02,Delivery Delay,Email,9.8,5.0,My order ORD107221 still hasn't arrived and it...
1,TIX100001,CUST003557,2025-08-29,Wrong Item Shipped,Phone,14.2,4.0,When will my refund for the returned item be p...
2,TIX100002,CUST002786,2024-10-31,Damaged Item,Social Media,2.1,4.0,Can you tell me the warranty period on this pr...
3,TIX100003,CUST003978,2026-06-24,Billing Issue,Social Media,16.2,5.0,"I can't log into my account, it says my passwo..."
4,TIX100004,CUST001008,2025-09-13,Damaged Item,Phone,8.8,5.0,"I can't log into my account, it says my passwo..."


In [ ]:
# Inspecting the Columns
print("Customers columns:")
print(customers.columns.tolist())

print("\nProducts columns:")
print(products.columns.tolist())

print("\nOrders columns:")
print(orders.columns.tolist())

print("\nOrder Items columns:")
print(order_items.columns.tolist())
print("\nSupport Tickets columns:")
print(support_tickets.columns.tolist())

Customers columns:
['customer_id', 'first_name', 'last_name', 'email', 'phone', 'gender', 'date_of_birth', 'signup_date', 'state', 'city', 'membership_tier', 'preferred_channel', 'marketing_opt_in', 'referral_source', 'Age', 'Age_Group', 'Order_Count', 'Total_Revenue', 'Customer_Segment', 'Revenue_ZScore', 'Revenue_Outlier', 'Ticket_Count', 'Average_Satisfaction']

Products columns:
['product_id', 'product_name', 'category', 'sub_category', 'brand', 'unit_cost', 'unit_price', 'launch_date', 'is_active', 'Price_Level']

Orders columns:
['order_id', 'customer_id', 'order_date', 'channel', 'order_status', 'payment_method', 'shipping_cost', 'discount_pct', 'total_amount']

Order Items columns:
['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price_at_purchase']

Support Tickets columns:
['ticket_id', 'customer_id', 'ticket_date', 'issue_category', 'channel', 'resolution_time_hours', 'satisfaction_score', 'ticket_text']


In [ ]:
# Checking Data types
print("========== CUSTOMERS DATA TYPES ==========")
print(customers.dtypes)

print("\n========== PRODUCTS DATA TYPES ==========")
print(products.dtypes)

print("\n========== ORDERS DATA TYPES ==========")
print(orders.dtypes)

print("\n========== ORDER ITEMS DATA TYPES ==========")
print(order_items.dtypes)
print("\n========== SUPPORT TICKETS DATA TYPES ==========")
print(support_tickets.dtypes)

========== CUSTOMERS DATA TYPES ==========
customer_id              object
first_name               object
last_name                object
email                    object
phone                    object
gender                   object
date_of_birth            object
signup_date              object
state                    object
city                     object
membership_tier          object
preferred_channel        object
marketing_opt_in         object
referral_source          object
Age                     float64
Age_Group                object
Order_Count             float64
Total_Revenue           float64
Customer_Segment         object
Revenue_ZScore          float64
Revenue_Outlier          object
Ticket_Count            float64
Average_Satisfaction    float64
dtype: object

========== PRODUCTS DATA TYPES ==========
product_id       object
product_name     object
category         object
sub_category     object
brand            object
unit_cost       float64
unit_price      floa

In [ ]:
# Checking missing values
print("========== CUSTOMERS MISSING VALUES ==========")
print(customers.isnull().sum())

print("\n========== PRODUCTS MISSING VALUES ==========")
print(products.isnull().sum())

print("\n========== ORDERS MISSING VALUES ==========")
print(orders.isnull().sum())

print("\n========== ORDER ITEMS MISSING VALUES ==========")
print(order_items.isnull().sum())
print("\n========== SUPPORT TICKETS MISSING VALUES ==========")
print(support_tickets.isnull().sum())

========== CUSTOMERS MISSING VALUES ==========
customer_id                0
first_name                 0
last_name                  0
email                    242
phone                    423
gender                     0
date_of_birth            181
signup_date              487
state                      0
city                     121
membership_tier            0
preferred_channel          0
marketing_opt_in        1191
referral_source         1705
Age                      354
Age_Group                  0
Order_Count                0
Total_Revenue              0
Customer_Segment           0
Revenue_ZScore             0
Revenue_Outlier            0
Ticket_Count               0
Average_Satisfaction    2843
dtype: int64

========== PRODUCTS MISSING VALUES ==========
product_id       0
product_name     0
category         0
sub_category    11
brand            0
unit_cost        0
unit_price      22
launch_date      0
is_active        0
Price_Level      0
dtype: int64

========== ORDERS MISS

In [ ]:
# Percentage of missing values
def missing_summary(df):
    summary = pd.DataFrame({
        "Missing Count": df.isnull().sum(),
        "Missing Percentage": (df.isnull().mean() * 100).round(2)
    })

    return summary[summary["Missing Count"] > 0].sort_values(
        "Missing Percentage",
        ascending=False
    )

In [ ]:
print("CUSTOMERS")
display(missing_summary(customers))

print("PRODUCTS")
display(missing_summary(products))

print("ORDERS")
display(missing_summary(orders))

print("ORDER ITEMS")
display(missing_summary(order_items))
print("SUPPORT TICKETS")
display(missing_summary(support_tickets))

CUSTOMERS


,Missing Count,Missing Percentage
Average_Satisfaction,2843,46.68
referral_source,1705,28.00
marketing_opt_in,1191,19.56
signup_date,487,8.00
phone,423,6.95
Age,354,5.81
email,242,3.97
date_of_birth,181,2.97
city,121,1.99


PRODUCTS


,Missing Count,Missing Percentage
unit_price,22,10.0
sub_category,11,5.0


ORDERS


,Missing Count,Missing Percentage
order_date,424,1.5


ORDER ITEMS


,Missing Count,Missing Percentage
unit_price_at_purchase,1469,2.0


SUPPORT TICKETS


,Missing Count,Missing Percentage
satisfaction_score,592,11.38


In [ ]:
# Checking duplicate Records
print("Customers duplicate rows:", customers.duplicated().sum())
print("Products duplicate rows:", products.duplicated().sum())
print("Orders duplicate rows:", orders.duplicated().sum())
print("Order Items duplicate rows:", order_items.duplicated().sum())
print("Support Tickets duplicate rows:", support_tickets.duplicated().sum())

Customers duplicate rows: 0
Products duplicate rows: 0
Orders duplicate rows: 25
Order Items duplicate rows: 40
Support Tickets duplicate rows: 0


In [ ]:
# Checking Categorical Values
def categorical_summary(df):
    categorical_columns = df.select_dtypes(
        include=["object", "category"]
    ).columns

    for column in categorical_columns:
        print(f"\n--- {column} ---")
        print(df[column].value_counts(dropna=False))

In [ ]:
print("========== CUSTOMERS ==========")
categorical_summary(customers)

print("\n========== PRODUCTS ==========")
categorical_summary(products)

print("\n========== ORDERS ==========")
categorical_summary(orders)

print("\n========== ORDER ITEMS ==========")
categorical_summary(order_items)
print("\n========== SUPPORT TICKETS ==========")
categorical_summary(support_tickets)

========== CUSTOMERS ==========

--- customer_id ---
customer_id
CUST002762    1
CUST002565    1
CUST004275    1
CUST003045    1
CUST005168    1
             ..
CUST005839    1
CUST004110    1
CUST000335    1
CUST003259    1
CUST000542    1
Name: count, Length: 6090, dtype: int64

--- first_name ---
first_name
Michael        122
Christopher    100
David           97
James           96
Jennifer        91
              ... 
Theodore         1
Geoffrey         1
Terrance         1
  Holly          1
  Sherry         1
Name: count, Length: 739, dtype: int64

--- last_name ---
last_name
Smith        152
Johnson      100
Williams      83
Jones         77
Brown         69
            ... 
Melendez       1
Hancock        1
Odonnell       1
Valentine      1
Crane          1
Name: count, Length: 933, dtype: int64

--- email ---
email
NaN                               242
joshua.grimes603@gmail.com          1
ashley.rios412@outlook.com          1
maxwell.bennett630@yahoo.com        1
kathryn.henr

In [ ]:
# Checking numerical variables for outliers
## Identifying numerical columns

print("Customers numerical columns:")
print(customers.select_dtypes(include="number").columns.tolist())

print("\nProducts numerical columns:")
print(products.select_dtypes(include="number").columns.tolist())

print("\nOrders numerical columns:")
print(orders.select_dtypes(include="number").columns.tolist())

print("\nOrder Items numerical columns:")
print(order_items.select_dtypes(include="number").columns.tolist())
print("\nSupport Tickets numerical columns:")
print(support_tickets.select_dtypes(include="number").columns.tolist())

Customers numerical columns:
['Age', 'Order_Count', 'Total_Revenue', 'Revenue_ZScore', 'Ticket_Count', 'Average_Satisfaction']

Products numerical columns:
['unit_cost', 'unit_price']

Orders numerical columns:
['shipping_cost', 'discount_pct', 'total_amount']

Order Items numerical columns:
['quantity', 'unit_price_at_purchase']

Support Tickets numerical columns:
['resolution_time_hours', 'satisfaction_score']


In [ ]:
# Calculating IQR outlier counts
def outlier_summary(df):
    numerical_columns = df.select_dtypes(include="number").columns

    results = []

    for column in numerical_columns:
        Q1 = df[column].quantile(0.25)
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1

        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        outliers = df[
            (df[column] < lower_bound) |
            (df[column] > upper_bound)
        ]

        results.append({
            "Column": column,
            "Q1": Q1,
            "Q3": Q3,
            "IQR": IQR,
            "Lower Bound": lower_bound,
            "Upper Bound": upper_bound,
            "Outlier Count": len(outliers)
        })

    return pd.DataFrame(results)

In [ ]:
print("========== CUSTOMERS OUTLIERS ==========")
display(outlier_summary(customers))

print("========== PRODUCTS OUTLIERS ==========")
display(outlier_summary(products))

print("========== ORDERS OUTLIERS ==========")
display(outlier_summary(orders))

print("========== ORDER ITEMS OUTLIERS ==========")
display(outlier_summary(order_items))
print("========== SUPPORT TICKETS OUTLIERS ==========")
display(outlier_summary(support_tickets))

========== CUSTOMERS OUTLIERS ==========


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,Age,35.000000,68.000000,33.000000,-14.500000,117.500000,0
1,Order_Count,3.000000,6.000000,3.000000,-1.500000,10.500000,52
2,Total_Revenue,4102.732500,9840.057500,5737.325000,-4503.255000,18446.045000,80
3,Revenue_ZScore,-0.699678,0.613356,1.313034,-2.669229,2.582907,80
4,Ticket_Count,0.000000,1.000000,1.000000,-1.500000,2.500000,328
5,Average_Satisfaction,3.000000,5.000000,2.000000,0.000000,8.000000,0


========== PRODUCTS OUTLIERS ==========


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,unit_cost,104.1300,309.5175,205.3875,-203.95125,617.59875,4
1,unit_price,180.1075,625.7900,445.6825,-488.41625,1294.31375,0


========== ORDERS OUTLIERS ==========


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,shipping_cost,0.00,7.99,7.99,-11.985,19.975,2
1,discount_pct,0.00,5.00,5.00,-7.500,12.500,3410
2,total_amount,671.23,2125.78,1454.55,-1510.595,4307.605,860


========== ORDER ITEMS OUTLIERS ==========


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,quantity,1.0000,2.0,1.0000,-0.50000,3.50000,3766
1,unit_price_at_purchase,174.6775,621.0,446.3225,-494.80625,1290.48375,0


========== SUPPORT TICKETS OUTLIERS ==========


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,resolution_time_hours,5.2,24.7,19.5,-24.05,53.95,270
1,satisfaction_score,3.0,5.0,2.0,0.00,8.00,0


In [ ]:
# ============================================================
# PHASE 6 — TARGETED INVESTIGATION BEFORE CLEANING
# ============================================================

print("=" * 70)
print("CUSTOMERS — DATE OF BIRTH")
print("=" * 70)

display(
    customers[
        ["customer_id", "date_of_birth", "Age"]
    ]
    .sort_values("date_of_birth")
    .head(20)
)

print("\nLatest DOB values:")
display(
    customers[
        ["customer_id", "date_of_birth", "Age"]
    ]
    .sort_values("date_of_birth", ascending=False)
    .head(20)
)


print("\n" + "=" * 70)
print("CUSTOMERS — REVENUE OUTLIERS")
print("=" * 70)

display(
    customers[
        ["customer_id", "Order_Count", "Total_Revenue",
         "Revenue_ZScore", "Revenue_Outlier"]
    ]
    .sort_values("Total_Revenue", ascending=False)
    .head(15)
)


print("\n" + "=" * 70)
print("PRODUCTS — UNIT COST OUTLIERS")
print("=" * 70)

display(
    products[
        ["product_id", "product_name", "unit_cost", "unit_price"]
    ]
    .sort_values("unit_cost", ascending=False)
    .head(15)
)


print("\n" + "=" * 70)
print("ORDERS — DISCOUNT VALUES")
print("=" * 70)

print("Discount statistics:")
print(orders["discount_pct"].describe())

print("\nUnique discount percentages:")
print(sorted(orders["discount_pct"].dropna().unique()))


print("\n" + "=" * 70)
print("ORDERS — HIGHEST TOTAL AMOUNTS")
print("=" * 70)

display(
    orders[
        ["order_id", "customer_id", "discount_pct",
         "shipping_cost", "total_amount"]
    ]
    .sort_values("total_amount", ascending=False)
    .head(15)
)


print("\n" + "=" * 70)
print("ORDER ITEMS — QUANTITY")
print("=" * 70)

print(order_items["quantity"].describe())

print("\nHighest quantities:")
display(
    order_items[
        ["order_item_id", "order_id", "product_id",
         "quantity", "unit_price_at_purchase"]
    ]
    .sort_values("quantity", ascending=False)
    .head(20)
)


print("\n" + "=" * 70)
print("ORDERS — DUPLICATE RECORDS")
print("=" * 70)

order_duplicates = orders[
    orders.duplicated(keep=False)
].sort_values("order_id")

display(order_duplicates)


print("\n" + "=" * 70)
print("ORDER ITEMS — DUPLICATE RECORDS")
print("=" * 70)

order_item_duplicates = order_items[
    order_items.duplicated(keep=False)
].sort_values("order_item_id")

display(order_item_duplicates)



CUSTOMERS — DATE OF BIRTH


,customer_id,date_of_birth,Age
3725,CUST002904,1899-01-01,NaN
281,CUST000564,1899-01-01,NaN
1379,CUST003108,1899-01-01,NaN
4407,CUST005845,1899-01-01,NaN
4739,CUST005105,1899-01-01,NaN
2965,CUST000821,1899-01-01,NaN
2877,CUST002675,1899-01-01,NaN
3617,CUST005989,1899-01-01,NaN
5055,CUST001825,1940-09-10,85.0
5540,CUST001997,1940-09-19,85.0



Latest DOB values:


,customer_id,date_of_birth,Age
3567,CUST000885,2031-05-04,NaN
1156,CUST004771,2031-05-04,NaN
5592,CUST003311,2031-05-04,NaN
1514,CUST002679,2031-05-04,NaN
317,CUST002417,2031-05-04,NaN
5799,CUST005029,2031-05-04,NaN
976,CUST004179,2031-05-04,NaN
4781,CUST000116,2010-08-24,NaN
32,CUST003555,2010-08-15,NaN
1469,CUST002690,2010-08-14,NaN



CUSTOMERS — REVENUE OUTLIERS


,customer_id,Order_Count,Total_Revenue,Revenue_ZScore,Revenue_Outlier
5311,CUST005678,14.0,35395.37,6.461899,Yes
950,CUST003680,13.0,28036.74,4.777816,Yes
3973,CUST000115,11.0,27460.18,4.645865,Yes
583,CUST000585,12.0,27017.52,4.544559,Yes
1321,CUST002212,13.0,26532.60,4.433581,Yes
1606,CUST003335,10.0,23960.29,3.844887,Yes
909,CUST001079,8.0,23620.82,3.767196,Yes
3012,CUST005624,7.0,23498.33,3.739163,Yes
1575,CUST005965,9.0,23495.75,3.738573,Yes
411,CUST003560,10.0,23216.87,3.674749,Yes



PRODUCTS — UNIT COST OUTLIERS


,product_id,product_name,unit_cost,unit_price
37,PROD1037,Pinecrest Smart Home Plus,1019.355,679.57
121,PROD1121,Vertex Team Sport,1019.340,679.56
27,PROD1027,Everwood Fitness Equipment Max,968.910,645.94
31,PROD1031,Lumen Cookware Essential,914.385,609.59
170,PROD1170,Cadence Accessories Max,599.085,399.39
195,PROD1195,Meridian Basics Headphones Plus,396.240,760.36
77,PROD1077,TrueForm Women's Clothing Max,394.320,655.79
113,PROD1113,Northfield Wellness X,389.310,735.66
25,PROD1025,Vertex Wellnes,387.960,955.43
17,PROD1017,Nova Skincare Lite,386.390,970.84



ORDERS — DISCOUNT VALUES
Discount statistics:
count    28297.000000
mean         3.485352
std          6.380998
min          0.000000
25%          0.000000
50%          0.000000
75%          5.000000
max         25.000000
Name: discount_pct, dtype: float64

Unique discount percentages:
[np.int64(0), np.int64(5), np.int64(10), np.int64(15), np.int64(20), np.int64(25)]

ORDERS — HIGHEST TOTAL AMOUNTS


,order_id,customer_id,discount_pct,shipping_cost,total_amount
2843,ORD122659,CUST001568,0,7.99,10536.77
7384,ORD109142,CUST004557,0,0.00,9937.54
18999,ORD104410,CUST004015,0,7.99,9092.24
8441,ORD118578,CUST006050,0,0.00,8905.20
16744,ORD127704,CUST005240,0,12.99,8826.77
14297,ORD115690,CUST004091,0,0.00,8688.53
10957,ORD106103,CUST001159,0,0.00,8651.32
15005,ORD127851,CUST002070,0,5.99,8648.08
801,ORD115492,CUST000882,0,7.99,8417.84
11374,ORD118514,CUST000743,0,5.99,8264.33



ORDER ITEMS — QUANTITY
count    73505.000000
mean         1.553663
std          0.870371
min         -4.000000
25%          1.000000
50%          1.000000
75%          2.000000
max          4.000000
Name: quantity, dtype: float64

Highest quantities:


,order_item_id,order_id,product_id,quantity,unit_price_at_purchase
73457,OI542064,ORD102068,PROD1203,4,200.73
25870,OI528049,ORD121678,PROD1064,4,13.42
73493,OI536853,ORD121403,PROD1144,4,251.61
12,OI517523,ORD103742,PROD1071,4,336.64
11201,OI529695,ORD114545,PROD1054,4,133.81
25936,OI547976,ORD124745,PROD1071,4,336.84
74,OI519981,ORD102840,PROD1090,4,203.86
25833,OI503743,ORD111016,PROD1015,4,363.09
11230,OI546778,ORD115453,PROD1012,4,453.77
11229,OI540430,ORD127242,PROD1163,4,487.66



ORDERS — DUPLICATE RECORDS


,order_id,customer_id,order_date,channel,order_status,payment_method,shipping_cost,discount_pct,total_amount
2873,ORD102482,CUST005630,2025-03-17,Mobile App,Completed,Credit Card,0.00,0,2852.97
10265,ORD102482,CUST005630,2025-03-17,Mobile App,Completed,Credit Card,0.00,0,2852.97
19831,ORD103359,CUST005402,2025-05-29,In-Store,Completed,Credit Card,0.00,0,420.05
15944,ORD103359,CUST005402,2025-05-29,In-Store,Completed,Credit Card,0.00,0,420.05
5795,ORD103787,CUST001807,2026-02-06,Mobile App,Completed,Debit Card,0.00,0,2633.64
23352,ORD103787,CUST001807,2026-02-06,Mobile App,Completed,Debit Card,0.00,0,2633.64
20404,ORD103914,CUST005431,2024-12-12,In-Store,Completed,Debit Card,5.99,5,2571.00
17094,ORD103914,CUST005431,2024-12-12,In-Store,Completed,Debit Card,5.99,5,2571.00
14105,ORD104413,CUST004015,2024-09-11,Mobile App,Completed,Store Credit,4.99,5,206.07
19106,ORD104413,CUST004015,2024-09-11,Mobile App,Completed,Store Credit,4.99,5,206.07



ORDER ITEMS — DUPLICATE RECORDS


,order_item_id,order_id,product_id,quantity,unit_price_at_purchase
41701,OI501045,ORD105556,PROD1058,2,503.85
46542,OI501045,ORD105556,PROD1058,2,503.85
17274,OI502291,ORD100604,PROD1049,2,20.33
51376,OI502291,ORD100604,PROD1049,2,20.33
40523,OI502812,ORD102369,PROD1213,2,457.96
...,...,...,...,...,...
62277,OI568084,ORD110101,PROD1192,1,725.06
50110,OI568833,ORD113939,PROD1006,1,530.95
51390,OI568833,ORD113939,PROD1006,1,530.95
53322,OI571744,ORD111844,PROD1137,4,416.30


**The above investigation gives me enough evidence to make several Phase 6 cleaning and transformation decisions confidently.**

 ***SOME OF THE FINDINGS INCLUDES***

1. Customers

Date of birth

1899-01-01 appears to be a placeholder/invalid DOB because those records also have missing Age.
2031-05-04 is a future date relative to the dataset period and is therefore invalid.
The normal DOBs appear reasonable.

Decision: Convert date_of_birth to datetime and replace clearly invalid placeholder/future dates with missing (NaT). We should not invent replacement dates.

Revenue outliers

The highest customer revenue is ₦35,395.37 with 14 orders.
The outlier customers have substantial order activity; they don't look like obvious data-entry errors.
Therefore, these are likely genuine high-value customers, which are especially important in a churn investigation.

Decision: Keep revenue outliers. Do not cap or remove them.

2. Products

The four unit_cost outliers are high, but importantly, they are still plausible positive monetary values.

However, there is an important issue in the data:

For some products, unit_cost is greater than unit_price.
For example, PROD1037 has unit cost 1019.355 but unit price 679.57.

That is potentially a business/data-quality issue, rather than simply a statistical outlier.

Decision for now: Don't remove these products merely because they are IQR outliers. We should preserve the observations and flag the cost-price inconsistencies for investigation rather than destroying potentially useful records.

3. Orders

The discount results are actually very good:

0, 5, 10, 15, 20, 25

All are sensible percentage values, and the maximum is only 25%.

Decision: Keep all discount values. Although IQR identified some as statistical outliers, they are valid business values and should not be removed.

The high total_amount values should also be retained. A large order is not automatically an error.

Decision: Keep transaction-value outliers unless we find evidence that they are impossible or incorrectly entered.

4. Order Items

This is the most important outlier problem we found.

quantity has:

Minimum = -4
Maximum = 4
Normal quantities appear to be 1–4.

A negative quantity is not a valid ordinary purchase quantity.

Decision: Negative quantities should be treated as invalid data, not as statistical outliers.

We should identify exactly how many negative quantities exist and decide whether they represent returns/cancellations or data-entry errors before removing them.

The positive quantity of 4 is perfectly reasonable, so we should not remove quantity = 4.

5. Duplicate records

We now have very strong evidence that the duplicate records are genuine exact duplicates.

For example:

ORD102482 appears twice with exactly the same customer, date, channel, status, payment method, shipping cost, discount and amount.

Similarly, the order-item duplicates have exactly the same:

order_item_id + order_id + product_id + quantity + unit_price_at_purchase.

Therefore:

Orders: Remove exact duplicate rows — 25 duplicate rows.

Order Items: Remove exact duplicate rows — 40 duplicate rows.

We should not remove rows simply because order_id repeats in order_items, because one order can legitimately contain multiple products/line items.

** IN CONCLUSION **

The data cleaning and transformation process was conducted systematically across the Customers, Products, Orders, and Order Items datasets. Decisions were based on the nature of each data-quality issue and its potential effect on the subsequent customer churn analysis. Missing values were retained where no reliable replacement could be determined, while selected variables were imputed where a defensible business-based estimate was available. Invalid dates, inconsistent categorical values, duplicate records, and inappropriate data types were addressed to improve data consistency and analytical reliability.

Statistical outliers were not automatically removed. Where extreme values represented plausible business activity, such as high customer revenue or high-value orders, they were retained because they may contain important information about customer value and churn behaviour. Conversely, logically invalid observations, such as negative quantities that could not be associated with returned orders, were treated as data-quality errors.

Exact duplicate records were removed from the Orders and Order Items datasets because they represented repeated copies of the same observations. However, repeated order IDs within the Order Items dataset were retained because one order may legitimately contain multiple line items.

Feature scaling was deliberately excluded from the cleaned datasets. Scaling will be applied during the modeling stage only where required by the selected machine-learning algorithms. This approach preserves the original business meaning of the cleaned data while ensuring that model-specific preprocessing is performed appropriately.



**DATA CLEANING AND TRANSFORMATION PROPAL**

In [21]:
# ============================================================
# PHASE 6 — DATA CLEANING & TRANSFORMATION
# Meridian Retail Group
# ============================================================

import pandas as pd
import numpy as np
import os

# ------------------------------------------------------------
# 1. MOUNT GOOGLE DRIVE
# ------------------------------------------------------------

from google.colab import drive
drive.mount('/content/drive')


# ------------------------------------------------------------
# 2. DEFINE FOLDERS
# ------------------------------------------------------------

processed_path = "/content/drive/MyDrive/Meridian Retail Group - Customer Churn Capstone /Data/Processed Data"
cleaned_path = "/content/drive/MyDrive/Meridian Retail Group - Customer Churn Capstone /Data/Cleaned Data"

# Create Cleaned Data folder if it does not exist
os.makedirs(cleaned_path, exist_ok=True)

print("Processed Data folder:")
print(processed_path)

print("\nCleaned Data folder:")
print(cleaned_path)




# ------------------------------------------------------------
# 3. LOAD PHASE 5 PROCESSED DATA
# ------------------------------------------------------------

customers = pd.read_csv(
    os.path.join(processed_path, "customers_processed.csv")
)

products = pd.read_csv(
    os.path.join(processed_path, "products_processed.csv")
)

orders = pd.read_csv(
    os.path.join(processed_path, "orders_processed.csv")
)

order_items = pd.read_csv(
    os.path.join(processed_path, "order_items_processed.csv")
)

support_tickets = pd.read_csv(
    os.path.join(processed_path, "support_tickets_processed.csv")
)

print("\nDatasets loaded successfully.")

print("\nDataset shapes:")
print("Customers:", customers.shape)
print("Products:", products.shape)
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Support Tickets:", support_tickets.shape)



# ============================================================
# 4. CUSTOMERS CLEANING
# ============================================================

print("\n" + "=" * 70)
print("CLEANING CUSTOMERS")
print("=" * 70)

# Make a copy so the original loaded dataframe is preserved
customers_clean = customers.copy()


# ------------------------------------------------------------
# 4.1 Convert date columns safely
# ------------------------------------------------------------

customers_clean["date_of_birth"] = pd.to_datetime(
    customers_clean["date_of_birth"],
    errors="coerce"
)

customers_clean["signup_date"] = pd.to_datetime(
    customers_clean["signup_date"],
    errors="coerce"
)


# ------------------------------------------------------------
# 4.2 Remove clearly invalid DOB values
# ------------------------------------------------------------

# 1899-01-01 appears to be a placeholder date
customers_clean.loc[
    customers_clean["date_of_birth"] == pd.Timestamp("1899-01-01"),
    "date_of_birth"
] = pd.NaT

# Future birth dates are invalid
customers_clean.loc[
    customers_clean["date_of_birth"] > pd.Timestamp("2026-09-09"),
    "date_of_birth"
] = pd.NaT


# ------------------------------------------------------------
# 4.3 Standardize gender
# ------------------------------------------------------------

customers_clean["gender"] = (
    customers_clean["gender"]
    .astype("string")
    .str.strip()
    .str.title()
)

customers_clean["gender"] = customers_clean["gender"].fillna("Unknown")


# ------------------------------------------------------------
# 4.4 Standardize membership tier
# ------------------------------------------------------------

customers_clean["membership_tier"] = (
    customers_clean["membership_tier"]
    .astype("string")
    .str.strip()
    .str.title()
)

customers_clean["membership_tier"] = (
    customers_clean["membership_tier"]
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 4.5 Standardize marketing opt-in
# ------------------------------------------------------------

marketing_map = {
    "Y": "Yes",
    "y": "Yes",
    "YES": "Yes",
    "Yes": "Yes",
    "1": "Yes",
    "1.0": "Yes",

    "N": "No",
    "n": "No",
    "NO": "No",
    "No": "No",
    "0": "No",
    "0.0": "No"
}

customers_clean["marketing_opt_in"] = (
    customers_clean["marketing_opt_in"]
    .astype("string")
    .str.strip()
    .replace(marketing_map)
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 4.6 Standardize state names
# ------------------------------------------------------------

state_map = {
    "NY": "New York",
    "AZ": "Arizona",
    "OH": "Ohio",
    "GA": "Georgia",
    "TX": "Texas",
    "MA": "Massachusetts",
    "PA": "Pennsylvania"
}

customers_clean["state"] = (
    customers_clean["state"]
    .astype("string")
    .str.strip()
    .replace(state_map)
)

customers_clean["state"] = customers_clean["state"].fillna("Unknown")


# ------------------------------------------------------------
# 4.7 Standardize referral source
# ------------------------------------------------------------

customers_clean["referral_source"] = (
    customers_clean["referral_source"]
    .astype("string")
    .str.strip()
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 4.8 Standardize city
# ------------------------------------------------------------

customers_clean["city"] = (
    customers_clean["city"]
    .astype("string")
    .str.strip()
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 4.9 Standardize preferred channel
# ------------------------------------------------------------

customers_clean["preferred_channel"] = (
    customers_clean["preferred_channel"]
    .astype("string")
    .str.strip()
    .str.title()
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 4.10 Handle missing satisfaction
# ------------------------------------------------------------

# Missing satisfaction is retained and flagged rather than
# blindly replacing it with an average.

customers_clean["Satisfaction_Missing"] = (
    customers_clean["Average_Satisfaction"].isna().astype(int)
)


# ------------------------------------------------------------
# 4.11 Remove exact duplicate customer records
# ------------------------------------------------------------

customers_clean = customers_clean.drop_duplicates()


# ============================================================
# 5. PRODUCTS CLEANING
# ============================================================

print("\n" + "=" * 70)
print("CLEANING PRODUCTS")
print("=" * 70)

products_clean = products.copy()


# ------------------------------------------------------------
# 5.1 Convert launch date
# ------------------------------------------------------------

products_clean["launch_date"] = pd.to_datetime(
    products_clean["launch_date"],
    errors="coerce"
)


# ------------------------------------------------------------
# 5.2 Standardize categorical values
# ------------------------------------------------------------

for col in ["category", "sub_category", "brand", "is_active", "Price_Level"]:

    products_clean[col] = (
        products_clean[col]
        .astype("string")
        .str.strip()
    )


# ------------------------------------------------------------
# 5.3 Handle missing sub-category
# ------------------------------------------------------------

products_clean["sub_category"] = (
    products_clean["sub_category"]
    .fillna("Unknown")
)


# ------------------------------------------------------------
# 5.4 Handle missing unit price
# ------------------------------------------------------------

# Use category-level median because products within the same
# category are more comparable than using one global median.

category_median_price = (
    products_clean.groupby("category")["unit_price"]
    .transform("median")
)

products_clean["unit_price"] = (
    products_clean["unit_price"]
    .fillna(category_median_price)
)


# ------------------------------------------------------------
# 5.5 Create a cost-price inconsistency flag
# ------------------------------------------------------------

products_clean["Cost_Greater_Than_Price"] = (
    products_clean["unit_cost"] > products_clean["unit_price"]
).astype(int)


# ------------------------------------------------------------
# 5.6 Remove exact duplicate product records
# ------------------------------------------------------------

products_clean = products_clean.drop_duplicates()


# ============================================================
# 6. ORDERS CLEANING
# ============================================================

print("\n" + "=" * 70)
print("CLEANING ORDERS")
print("=" * 70)

orders_clean = orders.copy()


# ------------------------------------------------------------
# 6.1 Convert order date
# ------------------------------------------------------------

orders_clean["order_date"] = pd.to_datetime(
    orders_clean["order_date"],
    errors="coerce"
)


# ------------------------------------------------------------
# 6.2 Standardize channel
# ------------------------------------------------------------

orders_clean["channel"] = (
    orders_clean["channel"]
    .astype("string")
    .str.strip()
    .str.title()
)


# ------------------------------------------------------------
# 6.3 Standardize order status
# ------------------------------------------------------------

orders_clean["order_status"] = (
    orders_clean["order_status"]
    .astype("string")
    .str.strip()
    .str.title()
)


# ------------------------------------------------------------
# 6.4 Standardize payment method
# ------------------------------------------------------------

orders_clean["payment_method"] = (
    orders_clean["payment_method"]
    .astype("string")
    .str.strip()
    .str.title()
)


# ------------------------------------------------------------
# 6.5 Ensure numeric columns have correct data types
# ------------------------------------------------------------

orders_clean["shipping_cost"] = pd.to_numeric(
    orders_clean["shipping_cost"],
    errors="coerce"
)

orders_clean["discount_pct"] = pd.to_numeric(
    orders_clean["discount_pct"],
    errors="coerce"
)

orders_clean["total_amount"] = pd.to_numeric(
    orders_clean["total_amount"],
    errors="coerce"
)


# ------------------------------------------------------------
# 6.6 Remove exact duplicate order records
# ------------------------------------------------------------

before_orders = len(orders_clean)

orders_clean = orders_clean.drop_duplicates()

after_orders = len(orders_clean)

print(f"Exact duplicate order rows removed: {before_orders - after_orders}")


# ------------------------------------------------------------
# 6.7 Keep missing order dates
# ------------------------------------------------------------

# We do NOT invent dates.
# Missing dates will remain NaT.


# ============================================================
# 7. ORDER ITEMS CLEANING
# ============================================================

print("\n" + "=" * 70)
print("CLEANING ORDER ITEMS")
print("=" * 70)

order_items_clean = order_items.copy()


# ------------------------------------------------------------
# 7.1 Ensure numeric columns have correct types
# ------------------------------------------------------------

order_items_clean["quantity"] = pd.to_numeric(
    order_items_clean["quantity"],
    errors="coerce"
)

order_items_clean["unit_price_at_purchase"] = pd.to_numeric(
    order_items_clean["unit_price_at_purchase"],
    errors="coerce"
)


# ------------------------------------------------------------
# 7.2 Identify negative quantities
# ------------------------------------------------------------

negative_quantity_count = (
    order_items_clean["quantity"] < 0
).sum()

print(
    f"Negative quantity records found: "
    f"{negative_quantity_count}"
)


# ------------------------------------------------------------
# 7.3 Check whether negative quantities are associated
#     with returned orders
# ------------------------------------------------------------

returned_orders = set(
    orders_clean.loc[
        orders_clean["order_status"] == "Returned",
        "order_id"
    ]
)

negative_quantity_rows = order_items_clean[
    order_items_clean["quantity"] < 0
].copy()

negative_quantity_rows["Associated_With_Return"] = (
    negative_quantity_rows["order_id"]
    .isin(returned_orders)
)

print("\nNegative quantities associated with returned orders:")
print(
    negative_quantity_rows[
        "Associated_With_Return"
    ].value_counts(dropna=False)
)


# ------------------------------------------------------------
# 7.4 Handle negative quantities
# ------------------------------------------------------------

# Negative quantities associated with returned orders may
# represent legitimate returns, so they are retained.
#
# Negative quantities NOT associated with returned orders
# are treated as invalid records and removed.

invalid_negative_rows = (
    (order_items_clean["quantity"] < 0)
    &
    (~order_items_clean["order_id"].isin(returned_orders))
)

removed_negative = invalid_negative_rows.sum()

order_items_clean = order_items_clean[
    ~invalid_negative_rows
].copy()

print(
    f"Invalid negative quantity rows removed: "
    f"{removed_negative}"
)


# ------------------------------------------------------------
# 7.5 Handle missing purchase prices
# ------------------------------------------------------------

# We use the product's cleaned unit price when the
# purchase price is missing.

product_price_lookup = (
    products_clean
    .set_index("product_id")["unit_price"]
)

missing_purchase_price = (
    order_items_clean["unit_price_at_purchase"].isna()
)

order_items_clean.loc[
    missing_purchase_price,
    "unit_price_at_purchase"
] = (
    order_items_clean.loc[
        missing_purchase_price,
        "product_id"
    ].map(product_price_lookup)
)


# ------------------------------------------------------------
# 7.6 Remove exact duplicate order-item records
# ------------------------------------------------------------

before_items = len(order_items_clean)

order_items_clean = order_items_clean.drop_duplicates()

after_items = len(order_items_clean)

print(
    f"Exact duplicate order-item rows removed: "
    f"{before_items - after_items}"
)


# ============================================================
# 8. SUPPORT TICKETS CLEANING
# ============================================================

print("\n" + "=" * 70)
print("CLEANING SUPPORT TICKETS")
print("=" * 70)

support_tickets_clean = support_tickets.copy()


# ------------------------------------------------------------
# 8.1 Display available columns
# ------------------------------------------------------------

print("\nSupport ticket columns:")
print(list(support_tickets_clean.columns))


# ------------------------------------------------------------
# 8.2 Remove exact duplicate ticket records
# ------------------------------------------------------------

before_tickets = len(support_tickets_clean)

support_tickets_clean = (
    support_tickets_clean
    .drop_duplicates()
    .copy()
)

after_tickets = len(support_tickets_clean)

print(
    f"\nExact duplicate support-ticket rows removed: "
    f"{before_tickets - after_tickets}"
)


# ------------------------------------------------------------
# 8.3 Standardize text columns
# ------------------------------------------------------------

# Clean all object/string columns by removing unnecessary
# leading/trailing spaces.

for col in support_tickets_clean.select_dtypes(
    include=["object", "string"]
).columns:

    support_tickets_clean[col] = (
        support_tickets_clean[col]
        .astype("string")
        .str.strip()
    )


# ------------------------------------------------------------
# 8.4 Standardize ticket status
# ------------------------------------------------------------

# If a status column exists, standardize its formatting.

status_candidates = [
    "status",
    "ticket_status",
    "Status",
    "Ticket_Status"
]

status_column = next(
    (
        col
        for col in status_candidates
        if col in support_tickets_clean.columns
    ),
    None
)

if status_column is not None:

    support_tickets_clean[status_column] = (
        support_tickets_clean[status_column]
        .astype("string")
        .str.strip()
        .str.title()
    )

    print(
        f"\nStatus column standardized: "
        f"{status_column}"
    )

else:

    print("\nNo ticket status column detected.")


# ------------------------------------------------------------
# 8.5 Convert satisfaction score to numeric
# ------------------------------------------------------------

satisfaction_candidates = [
    "satisfaction_score",
    "satisfaction",
    "Satisfaction_Score",
    "Satisfaction"
]

satisfaction_column = next(
    (
        col
        for col in satisfaction_candidates
        if col in support_tickets_clean.columns
    ),
    None
)

if satisfaction_column is not None:

    support_tickets_clean[satisfaction_column] = pd.to_numeric(
        support_tickets_clean[satisfaction_column],
        errors="coerce"
    )

    print(
        f"Satisfaction column converted to numeric: "
        f"{satisfaction_column}"
    )

else:

    print("No satisfaction column detected.")


# ------------------------------------------------------------
# 8.6 Convert resolution time to numeric
# ------------------------------------------------------------

resolution_candidates = [
    "resolution_time",
    "resolution_time_hours",
    "resolution_hours",
    "Resolution_Time",
    "Resolution_Time_Hours"
]

resolution_column = next(
    (
        col
        for col in resolution_candidates
        if col in support_tickets_clean.columns
    ),
    None
)

if resolution_column is not None:

    support_tickets_clean[resolution_column] = pd.to_numeric(
        support_tickets_clean[resolution_column],
        errors="coerce"
    )

    print(
        f"Resolution-time column converted to numeric: "
        f"{resolution_column}"
    )

else:

    print("No resolution-time column detected.")


# ------------------------------------------------------------
# 8.7 Convert date columns safely
# ------------------------------------------------------------

date_candidates = [
    "created_date",
    "ticket_date",
    "opened_date",
    "closed_date",
    "resolved_date",
    "created_at",
    "closed_at",
    "resolved_at",
    "Created_Date",
    "Ticket_Date",
    "Opened_Date",
    "Closed_Date",
    "Resolved_Date"
]

for col in date_candidates:

    if col in support_tickets_clean.columns:

        support_tickets_clean[col] = pd.to_datetime(
            support_tickets_clean[col],
            errors="coerce"
        )

        print(
            f"Date column converted: {col}"
        )


# ------------------------------------------------------------
# 8.8 Check negative resolution times
# ------------------------------------------------------------

if resolution_column is not None:

    negative_resolution = (
        support_tickets_clean[resolution_column] < 0
    ).sum()

    print(
        f"\nNegative resolution-time records found: "
        f"{negative_resolution}"
    )

    # Negative resolution time cannot be logically valid.
    # Therefore, convert these values to missing.

    support_tickets_clean.loc[
        support_tickets_clean[resolution_column] < 0,
        resolution_column
    ] = np.nan

    print(
        "Negative resolution times converted to missing values."
    )


# ------------------------------------------------------------
# 8.9 Validate satisfaction scores
# ------------------------------------------------------------

if satisfaction_column is not None:

    print("\nSatisfaction score summary:")
    print(
        support_tickets_clean[
            satisfaction_column
        ].describe()
    )

    # Satisfaction scores outside the normal 1–5 range
    # are treated as invalid.

    invalid_satisfaction = (
        (support_tickets_clean[satisfaction_column] < 1)
        |
        (support_tickets_clean[satisfaction_column] > 5)
    )

    invalid_satisfaction_count = (
        invalid_satisfaction.sum()
    )

    print(
        f"Invalid satisfaction scores found: "
        f"{invalid_satisfaction_count}"
    )

    support_tickets_clean.loc[
        invalid_satisfaction,
        satisfaction_column
    ] = np.nan

    print(
        "Invalid satisfaction scores converted to missing."
    )


# ------------------------------------------------------------
# 8.10 Validate customer IDs
# ------------------------------------------------------------

if "customer_id" in support_tickets_clean.columns:

    missing_customer_ids = (
        support_tickets_clean["customer_id"].isna()
    ).sum()

    print(
        f"\nMissing customer IDs: "
        f"{missing_customer_ids}"
    )


# ------------------------------------------------------------
# 8.11 Final support-ticket information
# ------------------------------------------------------------

print("\nSupport Tickets after cleaning:")
print(
    "Shape:",
    support_tickets_clean.shape
)

print("\nMissing values:")
print(
    support_tickets_clean.isna().sum()
)

print("\nData types:")
print(
    support_tickets_clean.dtypes
)



# ============================================================
# 9. FINAL VALIDATION
# ============================================================

print("\n" + "=" * 70)
print("FINAL VALIDATION")
print("=" * 70)


# ------------------------------------------------------------
# Customers
# ------------------------------------------------------------

print("\nCUSTOMERS")
print("Shape:", customers_clean.shape)
print("Exact duplicates:", customers_clean.duplicated().sum())
print("\nMissing values:")
print(customers_clean.isna().sum())


# ------------------------------------------------------------
# Products
# ------------------------------------------------------------

print("\nPRODUCTS")
print("Shape:", products_clean.shape)
print("Exact duplicates:", products_clean.duplicated().sum())
print("\nMissing values:")
print(products_clean.isna().sum())


# ------------------------------------------------------------
# Orders
# ------------------------------------------------------------

print("\nORDERS")
print("Shape:", orders_clean.shape)
print("Exact duplicates:", orders_clean.duplicated().sum())
print("\nMissing values:")
print(orders_clean.isna().sum())


# ------------------------------------------------------------
# Order Items
# ------------------------------------------------------------

print("\nORDER ITEMS")
print("Shape:", order_items_clean.shape)
print("Exact duplicates:", order_items_clean.duplicated().sum())
print("\nMissing values:")
print(order_items_clean.isna().sum())


# ------------------------------------------------------------
# Support Tickets
# ------------------------------------------------------------

print("\nSUPPORT TICKETS")
print(
    "Shape:",
    support_tickets_clean.shape
)

print(
    "Exact duplicates:",
    support_tickets_clean.duplicated().sum()
)

print("\nMissing values:")
print(
    support_tickets_clean.isna().sum()
)




# ============================================================
# 10. SAVE CLEANED DATASETS
# ============================================================

customers_clean.to_csv(
    os.path.join(
        cleaned_path,
        "customers_cleaned.csv"
    ),
    index=False
)

products_clean.to_csv(
    os.path.join(
        cleaned_path,
        "products_cleaned.csv"
    ),
    index=False
)

orders_clean.to_csv(
    os.path.join(
        cleaned_path,
        "orders_cleaned.csv"
    ),
    index=False
)

order_items_clean.to_csv(
    os.path.join(
        cleaned_path,
        "order_items_cleaned.csv"
    ),
    index=False
)

support_tickets_clean.to_csv(
    os.path.join(
        cleaned_path,
        "support_tickets_cleaned.csv"
    ),
    index=False
)


# ============================================================
# 11. CONFIRM FILES
# ============================================================

print("\n" + "=" * 70)
print("PHASE 6 COMPLETED — ALL FIVE DATASETS CLEANED")
print("=" * 70)

print("\nCleaned datasets saved to:")
print(cleaned_path)

print("\nFiles created:")

for file in sorted(os.listdir(cleaned_path)):
    print("✓", file)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Processed Data folder:
/content/drive/MyDrive/Meridian Retail Group - Customer Churn Capstone /Data/Processed Data

Cleaned Data folder:
/content/drive/MyDrive/Meridian Retail Group - Customer Churn Capstone /Data/Cleaned Data

Datasets loaded successfully.

Dataset shapes:
Customers: (6090, 23)
Products: (220, 10)
Orders: (28297, 9)
Order Items: (73505, 5)
Support Tickets: (5200, 8)

CLEANING CUSTOMERS

CLEANING PRODUCTS

CLEANING ORDERS
Exact duplicate order rows removed: 25

CLEANING ORDER ITEMS
Negative quantity records found: 18

Negative quantities associated with returned orders:
Associated_With_Return
False    17
True      1
Name: count, dtype: int64
Invalid negative quantity rows removed: 17
Exact duplicate order-item rows removed: 40

CLEANING SUPPORT TICKETS

Support ticket columns:
['ticket_id', 'customer_id', 'ticket_date', 'issue_category', 'cha